### PR TITLE
Verify `packSCMP`

### LINKED ISSUE #167: Make preconditions weaker/establish better framming for paths and DecodingLayers
Since PR #166, with the separation of the Mem predicates from the permissions to the underlying slice, it is possible to weaken the preconditions for methods in `Path` and `DecodingLayers` and establish stronger framing properties. We should look for opportunities for simplification in these groups of packages

### PR DESCRIPTION
**Pull request by jcp19**
~~This PR depends on the fix for https://github.com/scionproto/scion/issues/4482 being merged in SCION. I will continue it after that.~~

~~Should be merged after #313~~

TODO:
- [x] come up with a spec for packSCMP (https://github.com/viperproject/VerifiedSCION/pull/243)
- [x] verify it according to the given spec (https://github.com/viperproject/VerifiedSCION/pull/243)
- [x] Cherry pick https://github.com/scionproto/scion/pull/4483
- [x] Fix #167  
- [x] https://github.com/viperproject/VerifiedSCION/issues/297
- [x] remove all `@TODO()` that precede calls to it

Notice that this PR still does not verify `prepareSCMP` to completion. To address that method, we need to distinguish between SCION packets obtained by decoding a `[]byte`, or by manually constructing it in the `Mem` predicate. For that, we should introduce a boolean ghost field for that purpose. ~~As such, we should wait until PR https://github.com/viperproject/VerifiedSCION/pull/337 is merged before addressing `prepareSCMP`, if it is worth the hassle at all~~

~~To be continued after #347~~

### CODE REVIEW

**Comment by jcp19** [comment]
supersedes #245

**Comment by jcp19** [comment]
I will continue with this PR after https://github.com/scionproto/scion/issues/4482 is clarified

File: router/dataplane.go [hunk 1]
```diff
@@ -3266,17 +3417,25 @@ func (b *bfdSend) Send(bfd *layers.BFD) error {
 // @ requires  p.scionLayer.ValidPathMetaData(ub)
 // @ requires  sl.AbsSlice_Bytes(ub, 0, len(ub))
```
Comment (jcp19) [inline router/dataplane.go:3418]: Adapt according to #167

File: pkg/slayers/path/scion/base_spec.gobra [hunk 2]
```diff
@@ -39,9 +39,7 @@ pred (b *Base) Mem() {
 	acc(&b.PathMeta.SegLen[1]) &&
 	acc(&b.PathMeta.SegLen[2]) &&
 	0 <= b.NumINF && b.NumINF <= MaxINFs &&
-	// the program defines 64 as the maximum number of hops,
-	// but this does not seem to be enforced anywhere.
-	0 <= b.NumHops && // b.NumHops <= MaxHops &&
```
Comment (jcp19) [inline pkg/slayers/path/scion/base_spec.gobra:44]: Undo and uncomment `b.NumHops <= MaxHops`

Comment (jcp19) [reply pkg/slayers/path/scion/base_spec.gobra:44]: done in a later PR

**Review by Dspil — APPROVED** [review]

**Pull request merged**